### PR TITLE
[dashboard] fix onboarding modal show with workspaces exists

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

Go with origin proposal, onboarding Modal in `/workspaces` page will only show with

- User is onboarding user(not select IDE yet)
- No workspaces in `/workspaces` page

See also https://github.com/gitpod-io/gitpod/pull/9624#issuecomment-1112395133

## Related Issue(s)

Fixes #

## How to test

1. Go to `/workspaces` verify that the IDE selection modal is visible, do not interact with it, just close the browser tab
2. Launch a new workspace via `/#<repo_url>`
3. Go to `/workspaces` verify that the modal  **does not show**
4. Delete your workspace created by step 2
5. Go back to `/workspaces` and verify that the modal is showing again

## Release Notes

```release-note
NONE
```

### CODE REVIEW

File: components/dashboard/src/workspaces/Workspaces.tsx [hunk 1]
```diff
@@ -44,139 +45,147 @@ export default function () {
         })();
     }, [teams, location]);
 
+    useEffect(() => {
+        setIsEmpty(!activeWorkspaces.length && !inactiveWorkspaces.length && !workspaceModel?.searchTerm);
+    }, [activeWorkspaces, inactiveWorkspaces, workspaceModel]);
```
Comment (andreafalzetti) [inline components/dashboard/src/workspaces/Workspaces.tsx:50]: Would it be esier to read if we had:

```
const hasWorkspaces = activeWorkspaces.length || inactiveWorkspaces.length
```

and below, we would check:

```
{hasWorkspaces ||  workspaceModel?.searchTerm (
```

**Comment by andreafalzetti** [comment]
Thanks for raising the PR 🙏

**Comment by mustard-mh** [comment]
@andreafalzetti  Any idea what is `workspaceModel?.searchTerm`? Maybe it don't need to be a condition of `<SelectIDEModal>`

**Comment by andreafalzetti** [comment]
> @andreafalzetti Any idea what is `workspaceModel?.searchTerm`? Maybe it don't need to be a condition of `<SelectIDEModal>`

I think it's the state for this search bar:

Screenshot 2022-04-29 at 23 26 30

File: components/dashboard/src/workspaces/Workspaces.tsx (hunk 1 above)
Comment (mustard-mh) [reply components/dashboard/src/workspaces/Workspaces.tsx]: > I think it's the state for this search bar:

I haven't noticed this feature 😂 Updated.

I add a condition `isStartWorkspaceModalVisible`, in case toggle cmd+O to select repo cancel will go with select code (no idea why we need)

**Review by andreafalzetti — APPROVED** [review]
LGTM

Tested, works as expected! 🚀

**Comment by laushinka** [comment]
Taking a look! @andreafalzetti Should I run this `without-vm`?

**Comment by andreafalzetti** [comment]
> 

I think for testing it works either way. The reason I've used `without-vm` is because with the changes to preview environment I was having issues connecting to the db with mysql-cli since the new approach required to SSH into the vm and I didn't know.

**Comment by mustard-mh** [comment]
FYI. Access VM preview environment k3s:
- Switch to correct branch
- Exec `./dev/preview/install-k3s-kubeconfig.sh`

Then, all steps (exec \ port-forward etc) like before

 cc @andreafalzetti  @laushinka

**Comment by laushinka** [comment]
@mustard-mh I'm unable to access the preview env for some reason 🤔

**Comment by mustard-mh** [comment]
This PR has not committed for a long time, so platform team's job will delete this harvester(VM) preview environment every 30 mins 🥲 cc @laushinka @andreafalzetti

**Comment by mustard-mh** [comment]
I'll rebase it to origin/main to keep it fresh

**Comment by mustard-mh** [comment]
Werft job done 🎉 cc  @laushinka

**Comment by laushinka** [comment]
Tested according to the steps. In step 5 I don't see a modal - is this the expectation?

**Comment by andreafalzetti** [comment]
> Tested according to the steps. In step 5 I don't see a modal - is this the expectation?

Step 5 -> The modal should appear again, once you've deleted all workspaces. The concept is, if you've started a workspace, the modal should not be visible.

**Comment by laushinka** [comment]
> Step 5 -> The modal should appear again, once you've deleted all workspaces. 

@andreafalzetti I don't have any workspaces anymore and I don't see the modal reappear.

Screenshot 2022-05-06 at 13 17 59

**Comment by gtsiolis** [comment]
> The modal should appear again, once you've deleted all workspaces.

@andreafalzetti I could see the modal but is this intended?

From UX perspective:
- If a user has seen or interacted with the modal once they should _never_ be asked to select an editor again with the modal.
- Modals like these should be shown at least once either :a: when entering _any_ page in the dashboard or :b: when starting a workspace (see https://github.com/gitpod-io/gitpod/pull/9663)

**Comment by andreafalzetti** [comment]
> > Step 5 -> The modal should appear again, once you've deleted all workspaces.
> 
> @andreafalzetti I don't have any workspaces anymore and I don't see the modal reappear.
> 
> Screenshot 2022-05-06 at 13 17 59

You have a `defaultIde` in your user preferences; (I've just checked in the DB), this means that you should not see the modal.

The `defaultIde` is also saved if you close the modal via the "X" -> in this case it defaults to `code`

**Comment by laushinka** [comment]
> > > Step 5 -> The modal should appear again, once you've deleted all workspaces.
> > 
> > 
> > @andreafalzetti I don't have any workspaces anymore and I don't see the modal reappear.
> > Screenshot 2022-05-06 at 13 17 59
> 
> You have a `defaultIde` in your user preferences; (I've just checked in the DB), this means that you should not see the modal.

Ah, I see. Would that defaultIde have to be removed from the DB for me to be able to test it again? 🙈

File: components/dashboard/src/workspaces/Workspaces.tsx [hunk 2]
```diff
@@ -44,139 +45,147 @@ export default function () {
...
-                                        viewBox="0 0 16 16"
-                                        width="16"
-                                        height="16"
-                                    >
-                                        <path
-                                            fill="#A8A29E"
-                                            d="M6 2a4 4 0 100 8 4 4 0 000-8zM0 6a6 6 0 1110.89 3.477l4.817 4.816a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 010 6z"
-                                        />
-                                    </svg>
+            {workspaceModel?.initialized && (
```
[the remark this answers is not in the document]
Comment (mustard-mh) [reply components/dashboard/src/workspaces/Workspaces.tsx:58]: @laushinka  No, it's truly a model not modal😂. see this file https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/workspaces/workspace-model.ts

[the remark this answers is not in the document]
Comment (laushinka) [reply components/dashboard/src/workspaces/Workspaces.tsx:58]: LOL thanks for that

**Comment by andreafalzetti** [comment]
> However, this will no longer be necessary once we merge #9663, correct? I could be missing something but wondering why we are not merging #9663 instead. Maybe #9663 is far from being merged or the check we're adding in this PR needs to be in there anyway? 🤷

You have a good point @gtsiolis - For the long term, I cannot think of a current use-case where merging this - considering we'll soon merge #9663 - we add much value 🤔 

In the shorter term, it could help prevent seeing the modal after you've already used a workspace once, which might have looked odd.

@mustard-mh what's your view on this? Shall we continue with this PR or abandon it?

**Comment by mustard-mh** [comment]
Initially, this PR was created as a quick fix for this bug, as it was simple to modify and would be much sooner than the PR https://github.com/gitpod-io/gitpod/pull/9663 for review.

In reality, both PRs have been in the review state for a long time.

if https://github.com/gitpod-io/gitpod/pull/9663 can be merged first, we can close this one.

Edge case ([How to test section step 3](https://github.com/gitpod-io/gitpod/pull/9663)) will cause this issue too, but it's really edge case, we can ignore it

cc @andreafalzetti @gtsiolis
